pgml-docs/docs/blog/scaling-postgresml-to-one-million-requests-per-second.md

```diff
@@ -0,0 +1,252 @@
+---
+
+# Scaling PostgresML to 1 Million Requests per Second
+
+<p class="author">
+  <img width="54px" height="54px" src="/images/team/lev.jpg" alt="Author" />
+  Lev Kokotov<br/>
+  November 7, 2022
+</p>
+
+The question "Does it Scale?" has become somewhat of a meme in software engineering. There is a good reason for it though, because most businesses plan for success. If your app, online store, or SaaS takes off, you want to be sure that the system powering it can serve all your customers.
+
+At PostgresML, we are very concerned with scale. Our engineering background took us through scaling OLTP and OLAP Postgres to 100 TB+, so we're certain that Postgres scales, but could we scale machine learning alongside it?
+
+In this post, we'll discuss some challenges facing machine learning inference with PostgresML, and how we solved them to achieve an impressive **1 million XGBoost predictions per second** on commodity hardware.
+
+If you missed our previous post and are wondering why someone would combine machine learning and Postgres, take a look at our PostgresML vs. Python [benchmark](/blog/postgresml-is-8x-faster-than-python-http-microservices).
+
+
+## An Image Worth Four Thousand Words
+
+Our thesis, and the reason why we chose Postgres as our host for machine learning, is that scaling machine learning inference is very similar to scaling read queries in a typical database cluster.
+
+Inference speed varies based on the model complexity (e.g. `n_estimators` for XGBoost) and the size of the dataset (how many features the model uses), which is analogous to query complexity and table size in the database world. Scaling the latter is mostly a solved problem.
+
+### System Architecture
+
+If you're a Postgres enthusiast (or a database engineer), scaling Postgres may not be a secret to you, and you can jump straight to the [results](#results). For everyone else, here is a diagram showing the final state of our system:
+
+<center>
+![Scaling PostgresML](/images/illustrations/scaling-postgresml-3.svg) <br />
+_System Architecture_
+</center>
+
+| Component | Description |
+|-----------|-------------|
+| Clients | Regular Postgres clients |
+| ELB | [Elastic Network Load Balancer](https://aws.amazon.com/elasticloadbalancing/) |
+| PgCat | A Postgres [pooler](https://github.com/levkk/pgcat/) with built-in load balancing, failover, and sharding |
+| Replica | Regular Postgres [replicas](https://www.postgresql.org/docs/current/high-availability.html) |
+| Primary | Regular Postgres primary |
+
+
+Our architecture has four components that may need to scale up or down based on load:
+
+1. Clients
+2. Load balancer
+3. [PgCat](https://github.com/levkk/pgcat/) pooler
+4. Postgres replicas
+
+We intentionally don't discuss scaling the primary in this post, because sharding, which is the most effective way to do so, is a fascinating subject that deserves its own series of posts. Spoiler alert: we sharded Postgres without any problems.
+
+#### Clients
```

Comment:
I think the audience should be relatively aware of clients and the load balancer, and these sections can be condensed to keep the pacing up. PgCat is the potentially interesting callout for this section.